File: search_mode.go
package puyo2

import (
	"fmt"
	"strings"
)

type DedupMode int

const (
	DedupModeOff DedupMode = iota
	DedupModeSamePairOrder
	DedupModeState
	DedupModeStateMirror
)

func (m DedupMode) String() string {
	switch m {
	case DedupModeOff:
		return "off"
	case DedupModeSamePairOrder:
		return "same_pair_order"
	case DedupModeState:
		return "state"
	case DedupModeStateMirror:
		return "state_mirror"
	default:
		return fmt.Sprintf("unknown(%d)", int(m))
	}
}

func ParseDedupMode(v string) (DedupMode, error) {
	switch strings.TrimSpace(strings.ToLower(v)) {
	case "", "off":
		return DedupModeOff, nil
	case "same_pair_order":
		return DedupModeSamePairOrder, nil
	case "state":
		return DedupModeState, nil
	case "state_mirror":
		return DedupModeStateMirror, nil
	default:
		return DedupModeOff, fmt.Errorf("unknown dedup mode %q", v)
	}
}

type SimulatePolicy int

const (
	SimulatePolicyDetailAlways SimulatePolicy = iota
	SimulatePolicyFastIntermediate
	SimulatePolicyFastAlways
)

func (p SimulatePolicy) String() string {
	switch p {
	case SimulatePolicyDetailAlways:
		return "detail_always"
	case SimulatePolicyFastIntermediate:
		return "fast_intermediate"
	case SimulatePolicyFastAlways:
		return "fast_always"
	default:
		return fmt.Sprintf("unknown(%d)", int(p))
	}
}

func ParseSimulatePolicy(v string) (SimulatePolicy, error) {
	switch strings.TrimSpace(strings.ToLower(v)) {
	case "", "detail_always":
		return SimulatePolicyDetailAlways, nil
	case "fast_intermediate":
		return SimulatePolicyFastIntermediate, nil
	case "fast_always":
		return SimulatePolicyFastAlways, nil
	default:
		return SimulatePolicyDetailAlways, fmt.Errorf("unknown simulate policy %q", v)
	}
}

File: crates/puyo2/src/search_mode.rs
use std::{fmt, str::FromStr};

use thiserror::Error;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
pub enum DedupMode {
    #[default]
    Off,
    SamePairOrder,
    State,
    StateMirror,
}

impl fmt::Display for DedupMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let text = match self {
            Self::Off => "off",
            Self::SamePairOrder => "same_pair_order",
            Self::State => "state",
            Self::StateMirror => "state_mirror",
        };
        f.write_str(text)
    }
}

#[derive(Clone, Debug, Error, PartialEq, Eq)]
#[error("unknown dedup mode {0:?}")]
pub struct ParseDedupModeError(pub String);

impl FromStr for DedupMode {
    type Err = ParseDedupModeError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value.trim().to_ascii_lowercase().as_str() {
            "" | "off" => Ok(Self::Off),
            "same_pair_order" => Ok(Self::SamePairOrder),
            "state" => Ok(Self::State),
            "state_mirror" => Ok(Self::StateMirror),
            _ => Err(ParseDedupModeError(value.to_string())),
        }
    }
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash)]
pub enum SimulatePolicy {
    #[default]
    DetailAlways,
    FastIntermediate,
    FastAlways,
}

impl fmt::Display for SimulatePolicy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let text = match self {
            Self::DetailAlways => "detail_always",
            Self::FastIntermediate => "fast_intermediate",
            Self::FastAlways => "fast_always",
        };
        f.write_str(text)
    }
}

#[derive(Clone, Debug, Error, PartialEq, Eq)]
#[error("unknown simulate policy {0:?}")]
pub struct ParseSimulatePolicyError(pub String);

impl FromStr for SimulatePolicy {
    type Err = ParseSimulatePolicyError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value.trim().to_ascii_lowercase().as_str() {
            "" | "detail_always" => Ok(Self::DetailAlways),
            "fast_intermediate" => Ok(Self::FastIntermediate),
            "fast_always" => Ok(Self::FastAlways),
            _ => Err(ParseSimulatePolicyError(value.to_string())),
        }
    }
}
